compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp

```diff
@@ -120,13 +120,17 @@ TEST(TestRtsanInterceptors, VallocDiesWhenRealtime) {
   ExpectNonRealtimeSurvival(Func);
 }
 
-#if SANITIZER_INTERCEPT_ALIGNED_ALLOC
 TEST(TestRtsanInterceptors, AlignedAllocDiesWhenRealtime) {
-  auto Func = []() { EXPECT_NE(nullptr, aligned_alloc(16, 32)); };
-  ExpectRealtimeDeath(Func, "aligned_alloc");
-  ExpectNonRealtimeSurvival(Func);
-}
+#if SANITIZER_APPLE
+  if (__builtin_available(macOS 10.15, *)) {
```

Comment:
This is the only place in compiler-rt that __builtin_available is used, is there a more preferable way to do this? If I don't, I get a warning that suggests this:

```
 warning: 'aligned_alloc' is only available on macOS 10.15 or newer [-Wu
nguarded-availability-new]
...
_malloc.h:65:7: note: 'aligned_alloc' has been marked as being
introduced in macOS 10.15 here, but the deployment target is macOS 10.13.0
   65 | void *aligned_alloc(size_t __alignment, size_t __size) __result_use_check __alloc_align(1) __alloc_size(2) _MALLOC_TYPED(malloc_type_aligned_alloc, 2) __OSX_AVAILABLE(10.15)
 __IOS_AVAILABLE(13.0) __TVOS_AVAILABLE(13.0) __WATCHOS_AVAILABLE(6.0);
 ...
 note: enclose 'aligned_alloc' in a __builtin_available check to silence
 this warning
 ```

Reply:
In my catch-up reading about this issue I found [this](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Concepts/WeakLinking.html) documentation page, which suggests comparing the weak function pointer to `NULL` as a check. Not sure if this would be preferable here, but thought I'd drop it here just in case you hadn't seen it already 👍

Reply:
Great find on those docs, I'll spend some time digging through them.

Taking the very first naive crack at it, it appears to have similar errors (please enclose in __builtin_available).

That was repro'd applying this patch:

```
diff --git a/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp b/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
index fb227a1d1bc0..ad5c52fe7afc 100644
--- a/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
+++ b/compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp
@@ -120,9 +120,11 @@ TEST(TestRtsanInterceptors, VallocDiesWhenRealtime) {
   ExpectNonRealtimeSurvival(Func);
 }

+extern void* aligned_alloc(size_t, size_t) __attribute__((weak_import));
+
 TEST(TestRtsanInterceptors, AlignedAllocDiesWhenRealtime) {
 #if SANITIZER_APPLE
-  if (__builtin_available(macOS 10.15, *)) {
+  if (aligned_alloc != nullptr)
 #endif // SANITIZER_APPLE
     auto Func = []() { EXPECT_NE(nullptr, aligned_alloc(16, 32)); };
     ExpectRealtimeDeath(Func, "aligned_alloc");
```

Reply:
Also, doing some additional digging here, it seems like this __builtin_available is used in almost all subprojects of LLVM except compiler-rt, so there is reasonable support for it in the superproject at least.

Reply:
Nice, in that case it sounds like `__builtin_available` is the recommended way forward, then?

Reply:
All signs seem to be pointing that way. I definitely worry (since the `std::variant` debacle) about using something in compiler-rt that hasn't been used before, but seeing as it is only in a test file it seems less risky.

Also to be clear, this test passes on my mac that has `aligned_alloc` available, so this all works as intended :)